google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/readrows/ReadRowsResumptionStrategy.java

```diff
@@ -0,0 +1,153 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.stub.readrows;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.retrying.StreamResumptionStrategy;
+import com.google.bigtable.v2.ReadRowsRequest;
+import com.google.bigtable.v2.ReadRowsRequest.Builder;
+import com.google.bigtable.v2.RowRange;
+import com.google.bigtable.v2.RowSet;
+import com.google.cloud.bigtable.data.v2.internal.ByteStringComparator;
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.common.base.Preconditions;
+import com.google.protobuf.ByteString;
+
+/**
+ * An implementation of a {@link StreamResumptionStrategy} for merged rows. This class tracks the
+ * last complete row seen and upon retry can build a request to resume the stream from were it left
+ * off.
+ *
+ * <p>This class is considered an internal implementation detail and not meant to be used by
+ * applications.
+ */
+@InternalApi
+public class ReadRowsResumptionStrategy<RowT>
+    implements StreamResumptionStrategy<ReadRowsRequest, RowT> {
+  private final RowAdapter<RowT> rowAdapter;
+  private ByteString lastKey = ByteString.EMPTY;
+  private long numProcessed;
+
+  public ReadRowsResumptionStrategy(RowAdapter<RowT> rowAdapter) {
+    this.rowAdapter = rowAdapter;
+  }
+
+  @Override
+  public boolean canResume() {
+    return true;
+  }
+
+  @Override
+  public StreamResumptionStrategy<ReadRowsRequest, RowT> createNew() {
+    return new ReadRowsResumptionStrategy<>(rowAdapter);
+  }
+
+  @Override
+  public void onProgress(RowT response) {
+    // Last key can come from both the last processed row key and a synthetic row marker. The
+    // synthetic row marker is emitted when the server has read a lot of data that was filtered out.
+    // So it can trim the start of the scan, but does not contribute to the row limit.
+    lastKey = rowAdapter.getKey(response);
+    if (!rowAdapter.isScanMarkerRow(response)) {
+      // Only real rows count towards the rows limit.
+      numProcessed++;
+    }
+  }
+
+  @Override
+  public ReadRowsRequest getResumeRequest(ReadRowsRequest request) {
+    if (lastKey.isEmpty()) {
+      return request;
+    }
+    Builder builder = request.toBuilder();
+
+    if (request.getRowsLimit() > 0) {
+      Preconditions.checkState(
+          request.getRowsLimit() >= numProcessed,
+          "Detected too many responses for the current row limit during a retry.");
+      builder.setRowsLimit(request.getRowsLimit() - numProcessed);
+    }
+
+    // Reset rows to scan.
+    builder.clearRows();
+
+    RowSet.Builder newRowSet = RowSet.newBuilder();
+
+    // Special case: empty query implies full table scan
+    if (request.getRows().getRowKeysList().isEmpty()
+        && request.getRows().getRowRangesList().isEmpty()) {
+      newRowSet.addRowRanges(RowRange.newBuilder().setStartKeyOpen(lastKey).build());
```

Comment:
Please make sure we have a test to cover this code path (full table scan).

Reply:
done